src/environment-logs-provider.ts

```diff
@@ -0,0 +1,168 @@
+import * as vscode from "vscode";
+import { Environment } from "./env0-environments-provider";
+import { apiClient } from "./api-client";
+import stripAnsi from "strip-ansi";
+import { DeploymentStatus, DeploymentStepStatus } from "./types";
+import axios from "axios";
+
+const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
+const pollStepLogsInterval = 1000;
+
+export class EnvironmentLogsProvider {
+  private static environmentLogsOutputChannel: vscode.OutputChannel;
+  private readonly abortController = new AbortController();
+  private _isAborted = false;
+  private readonly stepsAlreadyLogged: string[] = [];
+  constructor(private readonly env: Environment) {
+    this.logDeployment(this.env.latestDeploymentLogId).catch((e) => {
+      if (!axios.isCancel(e)) {
+        throw e;
+      }
+    });
+  }
+
+  abort() {
+    this.abortController.abort();
+    this._isAborted = true;
+    EnvironmentLogsProvider.environmentLogsOutputChannel.clear();
+  }
+
+  get isAborted() {
+    return this._isAborted;
+  }
+
+  private log(value: string) {
+    if (!this.isAborted) {
+      const channel = EnvironmentLogsProvider.environmentLogsOutputChannel;
+      if (!channel) {
+        throw new Error(
+          "environment logs output channel used before initialized"
+        );
+      }
+      channel.appendLine(value);
+    }
+  }
+
+  private async logDeployment(deploymentId: string) {
+    let previousStatus;
+    EnvironmentLogsProvider.environmentLogsOutputChannel.show();
+
+    const deployment = await apiClient.getDeployment(
+      deploymentId,
+      this.abortController
+    );
+
+    if (deployment.status === DeploymentStatus.QUEUED) {
+      this.log(`Deployment is queued! Waiting for it to start...`);
+    }
+
+    while (!this.isAborted) {
+      const { type, status } = await apiClient.getDeployment(
+        deploymentId,
+        this.abortController
+      );
+
+      if (
+        status === DeploymentStatus.QUEUED &&
+        previousStatus === DeploymentStatus.QUEUED
+      ) {
+        this.log(
+          "Queued deployment is still waiting for earlier deployments to finish..."
+        );
+      }
+
+      if (
+        status === DeploymentStatus.IN_PROGRESS &&
+        previousStatus === DeploymentStatus.QUEUED
+      ) {
+        this.log(
+          `Deployment reached its turn! ${
+            type === "deploy" ? "Deploying" : "Destroying"
```

Comment:
can it not be a drift, task?

Reply:
asked about that in vscode channel

Reply:
as I understand, we shouldn't log anything that is destroy/deploy/resumes. is that handled in the code? where do we filter out the other types?

Reply:
we logging the latest deployment for environment when selected or when environment action triggered by the user.
As discussed in the channel, I updated the log 6aeaac2179df3364981749bb923738786fa4eef2